x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/shared/FrozenCacheService.java

```diff
@@ -759,36 +760,42 @@ private void throwAlreadyEvicted() {
-                            try {
-                                ensureOpen();
-                                final long start = gap.start();
-                                assert regionOwners[sharedBytesPos].get() == CacheFileRegion.this;
-                                writer.fillCacheRange(
-                                    fileChannel,
-                                    physicalStartOffset() + gap.start(),
-                                    gap.start(),
-                                    gap.end() - gap.start(),
-                                    progress -> gap.onProgress(start + progress)
-                                );
-                                writeCount.increment();
-                            } finally {
-                                decRef();
+                    try {
+                        executor.execute(new AbstractRunnable() {
+
+                            @Override
+                            protected void doRun() throws Exception {
+                                if (CacheFileRegion.this.tryIncRef() == false) {
+                                    throw new AlreadyClosedException("Cache file channel has been released and closed");
+                                }
+                                try {
+                                    ensureOpen();
+                                    final long start = gap.start();
+                                    assert regionOwners[sharedBytesPos].get() == CacheFileRegion.this;
+                                    writer.fillCacheRange(
+                                        fileChannel,
+                                        physicalStartOffset() + gap.start(),
+                                        gap.start(),
+                                        gap.end() - gap.start(),
+                                        progress -> gap.onProgress(start + progress)
+                                    );
+                                    writeCount.increment();
+                                } finally {
+                                    decRef();
+                                }
+                                gap.onCompletion();
                             }
-                            gap.onCompletion();
-                        }
 
-                        @Override
-                        public void onFailure(Exception e) {
-                            gap.onFailure(e);
-                        }
-                    });
+                            @Override
+                            public void onFailure(Exception e) {
+                                gap.onFailure(e);
+                            }
+                        });
+                    } catch (Exception e) {
+                        logger.error(() -> new ParameterizedMessage("unexpected exception when submitting task to fill gap [{}]", gap), e);
+                        assert false : e;
+                        gap.onFailure(e);
```

Comment:
I'd prefer we did this in the executor itself, as in https://github.com/elastic/elasticsearch/pull/85217 (passing the exception to `AbstractRunnable#onRejection` if possible). It looks wrong to expect `execute(AbstractRunnable)` to throw anything, and we assume it doesn't in so many other places.

Reply:
Hah yes what you said in https://github.com/elastic/elasticsearch/pull/85239/files?diff=unified&w=1#r832453126, I didn't see your comment until submitting mine.

Reply:
Yes, I'm just not sure when/how #85217 will be merged so I'd prefer to fix this here and later removed it if #85217 takes care of that.

Reply:
ok I see 👍

Reply:
I opened #85775 for this.